TCMManagement/BusinessLayer/AppointmentService.cs

```diff
@@ -1,85 +1,129 @@
-                return false;
-            }
-            context.Appointments.Remove(a);
-            SaveChanges();
-            return true;
-        }
-
-        public void MarkAsModified(Appointment item)
-        {
-            context.Entry(item).State = EntityState.Modified;
-        }
-
-        public int SaveChanges()
-        {
-            return context.SaveChanges();
-        }
-
-        public void Dispose()
-        {
-            context.Dispose();
-        }
-    }
+using System.Collections.Generic;
+using System.Linq;
+using TCMManagement.Models;
+using System.Data.Entity;
+using static TCMManagement.BusinessLayer.Constants;
+using TCMManagement.BusinessLayer;
+using System;
+using System.Web.Http.OData;
+
+namespace TCMManagement.BusinessLayer
+{
+    public class AppointmentService : IEntityServices<Appointment>
+    {
+        private TcmContext context;
+
+        public AppointmentService()
+        {
+            context = new TcmContext();
+        }
+
+        // NOTE: We might need to check the validity of this appointment.
+        public Appointment CreateItem(Appointment a)
+        {
+            // check the start time of the new appointment
+            var conflictList = context.Appointments
+                                .Where(ea => ea.PatientId == a.PatientId)
+                                .Where(ea => ea.PersonId == a.PersonId)
+                                .ToList();
+            foreach (var conflictAppointment in conflictList)
+            {
+                // any time overlap will cause conflict
+                if ((conflictAppointment.TimeStart >= a.TimeStart && conflictAppointment.TimeStart <= a.TimeEnd)
+                 || (conflictAppointment.TimeEnd >= a.TimeStart && conflictAppointment.TimeEnd <= a.TimeEnd)
+                 || (a.TimeStart >= conflictAppointment.TimeStart && a.TimeStart <= conflictAppointment.TimeEnd)
+                 || (a.TimeEnd >= conflictAppointment.TimeStart && a.TimeEnd <= conflictAppointment.TimeEnd))
+                {
+                    return null;
+                }
+            }
+            // no conflict, add the appointment
+            context.Appointments.Add(a);
+            SaveChanges();
+            return context.Appointments.ToList().Last();
+        }
+
+        public IEnumerable<Appointment> GetItems(IEnumerable<KeyValuePair<string, string>> queryParams = null)
+        {
+            if(!Utils.IsNullOrEmpty(queryParams)){
+                KeyValuePair<string, string> p = queryParams.FirstOrDefault();
+                bool isPatient = p.Key == "Patient";
+                int id = Int32.Parse(p.Value);
+
+                if(isPatient)
+                    return context.Appointments.Where(a => a.PatientId == id ).ToList();
+                else
+                {
+                    List<KeyValuePair<string, string>> queryList = queryParams.ToList();
+                    DateTime timeStart = DateTime.Parse(queryList[1].Value);

```

Comment:
I think we might need to be able to handle the case when user doesn't supply search params for timeStart and timeEnd.
-- just to make it more robust. 
Otherwise it will be very easy to crash here.

Reply:
Will add the case where only personId is available. 
But I don't think we need to cover the case while only timeStart or timeEnd is available.

Reply:
yeah, you are right. we don't need that.
PersonID or PatientID should be minimum requirement.

Reply:
Will add another check to make sure timeStart < timeEnd